synapse/rest/admin/rooms.py

```diff
@@ -340,3 +343,136 @@ async def on_POST(self, request, room_identifier):
         )
 
         return 200, {"room_id": room_id}
+
+
+class MakeRoomAdminRoomServlet(RestServlet):
+    """Allows a server admin to get power in a room if a local user has power in
+    a room. Will also invite the user if they're not in the room and its a
+    private room. Can specify another user (rather than the admin user) to be
+    granted power, e.g.:
+
+        POST /_synapse/admin/v1/make_room_admin/<room_id_or_alias>
+        {
+            "user_id": "@foo:example.com"
+        }
+    """
+
+    PATTERNS = admin_patterns("/make_room_admin/(?P<room_identifier>[^/]*)")
+
+    def __init__(self, hs: "HomeServer"):
+        self.hs = hs
+        self.auth = hs.get_auth()
+        self.room_member_handler = hs.get_room_member_handler()
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.state_handler = hs.get_state_handler()
+        self.is_mine_id = hs.is_mine_id
+
+    async def on_POST(self, request, room_identifier):
+        requester = await self.auth.get_user_by_req(request)
+        await assert_user_is_admin(self.auth, requester.user)
+        content = parse_json_object_from_request(request, allow_empty_body=True)
+
+        # Resolve to a room ID, if necessary.
+        if RoomID.is_valid(room_identifier):
+            room_id = room_identifier
+        elif RoomAlias.is_valid(room_identifier):
+            room_alias = RoomAlias.from_string(room_identifier)
+            room_id, _ = await self.room_member_handler.lookup_room_alias(room_alias)
+            room_id = room_id.to_string()
+        else:
+            raise SynapseError(
+                400, "%s was not legal room ID or room alias" % (room_identifier,)
+            )
+
+        # Which user to grant room admin rights to.
+        user_to_add = content.get("user_id", requester.user.to_string())
```

Comment:
Would it make sense to ensure this is a local user? Maybe it doesn't matter.

Reply:
I think it's fine to set it as a remote user, a bit weird but :shrug: